src/utils.js

```diff
@@ -379,17 +380,82 @@ function findSchemaDefinition($ref, definitions = {}) {
   throw new Error(`Could not find a definition for ${$ref}.`);
 }
 
-export function retrieveSchema(schema, definitions = {}) {
-  // No $ref attribute found, returning the original schema.
-  if (!schema.hasOwnProperty("$ref")) {
+export function retrieveSchema(schema, definitions = {}, formData = {}) {
+  if (schema.hasOwnProperty("$ref")) {
+    // Retrieve the referenced schema definition.
+    const $refSchema = findSchemaDefinition(schema.$ref, definitions);
+    // Drop the $ref property of the source schema.
+    const { $ref, ...localSchema } = schema;
+    // Update referenced schema definition with local schema properties.
+    return retrieveSchema(
+      { ...$refSchema, ...localSchema },
+      definitions,
+      formData
+    );
+  } else if (schema.hasOwnProperty("dependencies")) {
+    // Drop the dependencies from the source schema.
+    let { dependencies, ...localSchema } = schema;
+    // Process dependencies updating the local schema properties as appropriate.
+    for (const key of Object.keys(dependencies)) {
+      // Skip this dependency if its trigger property is not present.
+      if (!formData[key]) {
+        // fixme: a falsey check may not be appropriate here, I'm not sure
+        continue;
```

Comment:
AFAICR it depends if the field value has been edited already or not; by default a schema string field value is set to undefined (property is not preset), but as soon as users manipulate the widget, eg. typing then erasing the entered value, it's set to an empty string. So yeah, it might be a little tricky.